### DISCUSSION TITLE
classes/class_array

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#251] [User-contributed notes]
# classes/class_array

A built-in data structure that holds a sequence of elements. Description: An array data structure that can contain a sequence of elements of any Variant type. Elements are accessed by a numerical i...

https://docs.godotengine.org/en/stable/classes/class_array.html

**Comment by whogben** [comment]
The documentation example of sort_custom's behavior is inaccurate - it reads (in part):
```gdscript
# Sort descending, using a lambda function.
my_items.sort_custom(func(a, b): return a[0] > b[0])
print(my_items) # Prints [["Apple", 9], ["Tomato", 5], ["Rice", 4]]
```

It should instead read:
```gdscript
# Sort descending, using a lambda function.
my_items.sort_custom(func(a, b): return a[0] > b[0])
print(my_items) # Prints [["Tomato", 5], ["Rice", 4], ["Apple", 9]]
```

Tested on Godot 4.3

**Reply by Calinou** [reply]
I can't reproduce this. Here, I get the following on 4.3.stable when I copy-paste the code sample and run it:

```gdscript
# Sort ascending
[["Rice", 4], ["Tomato", 5], ["Apple", 9]]

# Sort descending
[["Apple", 9], ["Tomato", 5], ["Rice", 4]]
```

**Reply by timothyqiu** [reply]
The indices used in the lambda are different.

It has been fixed in https://github.com/godotengine/godot/pull/96118

**Comment by GersonFeDutra** [comment]
The example in the slice method's description is inaccurate, on reverse slicing:

```gdscript
var letters = ["A", "B", "C", "D", "E", "F"]

print(letters.slice(2, -2)) # Prints ["C", "D"]
print(letters.slice(4, 1, -1)) # Prints ["E", "D", "C"]
```

Correct output:
```gdscript
print(letters.slice(2, -2)) # Prints ["C", "D", "E"]
print(letters.slice(4, 1, -1)) # Prints ["E", "D", "C", "B"]
```

It's actually **inclusive** with the negative step.

**Reply by timothyqiu** [reply]
You are probably using the wrong version. It's inclusive in 3.x, but exclusive in 4.x.

**Reply by GersonFeDutra** [reply]
You are right indeed. I was testing on [gdscript online](https://gdscript-online.github.io/) and didn't saw the version.

I was trying to get a full reverse slice with:

```gdscript
print(letters.slice(4, -1, -1)) # Prints []
```

and
```gdscript
print(letters.slice(4, 0, -1)) # Prints ["E", "D", "C", "B"]
```


Given that -1 its a count from end, I was able to get the full reverse array with the *`ackward`*:
```gdscript
var index: int = 4
print(letters.slice(index - letters.size(), -(letters.size() + 1), -1)) # Prints ["E", "D", "C", "B", "A"]
```

**Comment by akumagamo** [comment]
Just stumbled upon this, the docs for the `sort_custom` are incorrect, it states  _"The function should return **true** if the first element should be moved **before** the second one, otherwise it should return false."_ and in the editor help it says"...**behind**...", what produces the correct result.

**Comment by RedpingDev** [comment]
In the Godot 4 game engine documentation, the description for the array append function has a mistake in the variable name. It should be "numbers" but it is incorrectly written as "nums."

```
var numbers = [1, 2, 3]
var extra = [4, 5, 6]
numbers.append_array(extra)
print(nums) # Prints [1, 2, 3, 4, 5, 6]
```

**Reply by Calinou** [reply]
This was fixed in https://github.com/godotengine/godot/pull/96033, but it's not part of the stable class reference yet. You should see it being fixed on `/latest` already.

**Comment by BaconEggsRL** [comment]
Note: Arrays marked const in GDScript will give a constant pointer to the given array, rather than making the array itself a constant.

Immutable arrays are apparently not a thing.

https://godotforums.org/d/23252-array-typed-as-const-is-not-constant

**Reply by PeraBajaDev** [reply]
Use make_read_only() for inmuttable arrays

**Comment by DevSavvySerge** [comment]
About sort_custom (https://docs.godotengine.org/en/stable/classes/class_array.html#class-array-method-sort-custom) how are we supposed to deal when "a" & "b" share the same priority (for instance, when a & b are equals)?

**Reply by timothyqiu** [reply]
> otherwise it should return `false`

**Comment by astery** [comment]
A common way to remove an item from an array in a loop:

```gdscript
var i: int = 0
while i < array.size():
  if should remove:
    # remove_at() resizes an array
    array.remove_at(i)
  else:
    i += 1
```

**Reply by timothyqiu** [reply]
Going from the end is a bit simpler:

```gdscript
for i in range(array.size() - 1, -1, -1):
	if should remove:
		array.remove_at(i)
```

**Comment by xsellier** [comment]
It'd be great to add another example for the `find`custom`, like this one:
```gdscript
func is_specific_number(number, expected):
    return number == expected

func _ready():
    print([1, 3, 4, 7].find_custom(is_specific_number.bind(4))) # Prints 2
```

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/115659

**Comment by BinaryGleam** [comment]
In the code in sort custom, there's a bit of bloating (maybe, I don't know. I'm commenting to see people opinion about this)

The example code starts with:
```gdscript
func sort_ascending(a, b):
	if a[1] < b[1]:
		return true
	return false
```
where it could be simplified to:
```gdscript
func sort_ascending(a, b):
	return a[1] < b[1]
```
Right? Or there's something I'm not seeing 😅

**Reply by Calinou** [reply]
It's true that it can be written this way, but I suppose the code sample is intentionally written in a longer form to make it more obvious that the function should return `true` or `false` according to the description.

**Reply by BinaryGleam** [reply]
I understand that but any bool operation implicitly returns true or false... It could also be changed to
```gdscript
func sort_ascending(a, b) -> bool:
	return a[1] < b[1]
```
And then it would be less bloated and still be explicit. If it has to return a bool I think it would be nice that the function definition forces it for the user too. Still open to dialogue about all this

**Reply by Calinou** [reply]
We currently don't use type hints in the class reference. This may change in the future though:

- https://github.com/godotengine/godot-docs/issues/6295

**Comment by Harizio2** [comment]
If methods with Array[Type] parameters can be called with a plain array initializer can depend on the context.
Array initializers in method calls get implicitly typed if the method owners type is known.
Autoload type is inferred for plain scripts, but not scenes.

Methods with PackedTypeArray parameters can be called more flexibly, but presumably create a copy when called with regular arrays.

```gdscript
class_name Foo extends Node

func _ready() -> void:
	# direct
	test(["bar"]) # works
	
	# indirect 
	(self as Node).test(["bar"]) # ERROR
	call("test",["bar"]) # ERROR
	call("test",["bar"] as Array[String]) # works

	# autoload
	AutoLoadFooScript.test(["bar"]) # works
	AutoLoadFooScene.test(["bar"]) # ERROR
	(AutoLoadFooScene as Foo).test(["bar"]) # works
	pass
	
func test(arr: Array[String])->void:
	prints(arr)
```
